File: docs/site/versioned_examples/version-21.09.0/cpp/pipeline.cpp
#include "asapo/asapo_producer.h"
#include "asapo/asapo_consumer.h"
#include <iostream>

void ProcessAfterSend(asapo::RequestCallbackPayload payload, asapo::Error err) {
    if (err && err != asapo::ProducerErrorTemplates::kServerWarning) {
        std::cerr << "error during send: " << err << std::endl;
        return;
    } else if (err) {
        std::cout << "warning during send: " << err << std::endl;
    } else {
        std::cout << "successfuly send " << payload.original_header.Json() << std::endl;
        return;
    }
}

void exit_if_error(std::string error_string, const asapo::Error& err) {
    if (err) {
        std::cerr << error_string << err << std::endl;
        exit(EXIT_FAILURE);
    }
}

int main(int argc, char* argv[]) {
    asapo::Error err;

    auto endpoint = "localhost:8400";
    auto beamtime = "asapo_test";

    auto token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJl"
                 "eHAiOjk1NzE3MTAyMTYsImp0aSI6ImMzaXFhbGpmN"
                 "DNhbGZwOHJua20wIiwic3ViIjoiYnRfYXNhcG9fdG"
                 "VzdCIsIkV4dHJhQ2xhaW1zIjp7IkFjY2Vzc1R5cGV"
                 "zIjpbIndyaXRlIiwicmVhZCJdfX0.dkWupPO-ysI4"
                 "t-jtWiaElAzDyJF6T7hu_Wz_Au54mYU";

    auto path_to_files = "/var/tmp/asapo/global_shared/data/test_facility/gpfs/test/2019/data/asapo_test";

    auto credentials = asapo::SourceCredentials{asapo::SourceType::kProcessed, beamtime, "", "test_source", token};

    auto producer = asapo::Producer::Create(endpoint, 1, asapo::RequestHandlerType::kTcp, credentials, 60000, &err);
    exit_if_error("Cannot start producer", err);
    auto consumer = asapo::ConsumerFactory::CreateConsumer(endpoint, path_to_files, true, credentials, &err);
    exit_if_error("Cannot start consumer", err);
    consumer->SetTimeout(5000);
    auto group_id = consumer->GenerateNewGroupId(&err);
    exit_if_error("Cannot create group id", err);

    // put the processed message into the new stream
    auto pipelined_stream_name = "pipelined";

    asapo::MessageMeta mm;
    asapo::MessageData data;

    do {
        // we expect the message to be in the 'default' stream already
        err = consumer->GetNext(group_id, &mm, &data, "default");

        if (err && err == asapo::ConsumerErrorTemplates::kStreamFinished) {
            std::cout << "stream finished" << std::endl;
            break;
        }

        if (err && err == asapo::ConsumerErrorTemplates::kEndOfStream) {
            std::cout << "stream ended" << std::endl;
            break;
        }
        exit_if_error("Cannot get next record", err);

        // work on our data
        auto processed_string = std::string(reinterpret_cast<char const*>(data.get())) + " processed";
        auto send_size = processed_string.size() + 1;
        auto buffer = asapo::MessageData(new uint8_t[send_size]);
        memcpy(buffer.get(), processed_string.c_str(), send_size);

        // you may use the same filename, if you want to rewrite the source file. This will result in warning, but it is a valid usecase
        asapo::MessageHeader message_header{mm.id, send_size, std::string("processed/test_file_") + std::to_string(mm.id)};
        err = producer->Send(message_header, std::move(buffer), asapo::kDefaultIngestMode, pipelined_stream_name,
                             &ProcessAfterSend);
        exit_if_error("Cannot send message", err);
    } while (1);


    err = producer->WaitRequestsFinished(2000);
    exit_if_error("Producer exit on timeout", err);

    // the meta from the last iteration corresponds to the last message
    auto last_id = mm.id;

    err = producer->SendStreamFinishedFlag("pipelined", last_id, "", &ProcessAfterSend);
    exit_if_error("Cannot finish stream", err);

    // you can remove the source stream if you do not need it anymore
    err = consumer->DeleteStream("default", asapo::DeleteStreamOptions{true, true});

    return EXIT_SUCCESS;
}

File: docs/site/versioned_examples/version-22.03.0/python/pipeline.py
import asapo_consumer
import asapo_producer

def callback(payload,err):
    if err is not None and not isinstance(err, asapo_producer.AsapoServerWarning):
        print("could not send: ",payload,err)
    elif err is not None:
        print("sent with warning: ",payload,err)
    else:
        print("successfuly sent: ",payload)

endpoint = "localhost:8400"
beamtime = "asapo_test"

token = str("eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e"
"yJleHAiOjk1NzE3MTAyMTYsImp0aSI6ImMzaXFhbGpmNDNhbGZ"
"wOHJua20wIiwic3ViIjoiYnRfYXNhcG9fdGVzdCIsIkV4dHJhQ"
"2xhaW1zIjp7IkFjY2Vzc1R5cGVzIjpbIndyaXRlIiwicmVhZCJ"
"dfX0.dkWupPO-ysI4t-jtWiaElAzDyJF6T7hu_Wz_Au54mYU")

path_to_files = "/var/tmp/asapo/global_shared/data/test_facility/gpfs/test/2019/data/asapo_test"

consumer = asapo_consumer.create_consumer(endpoint, path_to_files, True, beamtime, "test_source", token, 5000)

producer = asapo_producer.create_producer(endpoint, 'processed', beamtime, 'auto', 'test_source', '', 1, 60000)

group_id = consumer.generate_group_id()
# pipeline snippet_start
# put the processed message into the new stream
pipelined_stream_name = 'pipelined'

try:
    while True:
        # we expect the message to be in the 'default' stream already
        data, meta = consumer.get_next(group_id, meta_only = False)
        message_id = meta['_id']
        
        # work on our data
        text_data = data.tobytes().decode("utf-8")
        pipelined_message = (text_data + ' processed').encode()
        
        # you may use the same filename, if you want to rewrite the source file. This will result in warning, but it is a valid usecase
        producer.send(message_id, "processed/test_file_" + message_id, pipelined_message, pipelined_stream_name, callback = callback)
        

except asapo_consumer.AsapoStreamFinishedError:
    print('stream finished')
        
except asapo_consumer.AsapoEndOfStreamError:
    print('stream ended')
# pipeline snippet_end
producer.wait_requests_finished(2000)

# finish snippet_start
# the meta from the last iteration corresponds to the last message
last_id = meta['_id']

producer.send_stream_finished_flag("pipelined", last_id)
# finish snippet_end

# you can remove the source stream if you do not need it anymore
consumer.delete_stream(stream = 'default', error_on_not_exist = True)
